r3bgen/R3BParticleGenerator.cxx

```diff
@@ -0,0 +1,137 @@
+ *                    copied verbatim in the file "LICENSE".                  *
+ *                                                                            *
+ * In applying this license GSI does not waive the privileges and immunities  *
+ * granted to it by virtue of its status as an Intergovernmental Organization *
+ * or submit itself to any jurisdiction.                                      *
+ ******************************************************************************/
+
+#include "R3BParticleGenerator.h"
+#include "R3BDistribution2D.h"
+#include "R3BDistribution3D.h"
+
+#include "FairIon.h"
+#include "FairLogger.h"
+#include "FairParticle.h"
+#include "FairPrimaryGenerator.h"
+#include "FairRunSim.h"
+
+#include "TDatabasePDG.h"
+#include "TMath.h"
+#include "TObjArray.h"
+#include "TParticle.h"
+#include "TRandom.h"
+#include "TVector3.h"
+
+#include <iostream>
+
+using std::cout;
+using std::endl;
+
+constexpr auto AMU = 0.931494028;      // Gev/c**2
+constexpr auto ProtonMass = 0.938272;  // GeV
+constexpr auto NeutronMass = 0.939565; // GeV
+
+R3BParticleGenerator::R3BParticleGenerator()
+    : fVertex_cm(R3BDistribution3D::Delta(0, 0, 0))
+    , fSpread_mRad({ R3BDistribution1D::Delta(0), R3BDistribution1D::Flat(0, 2 * TMath::Pi() * 1e3) })
+    , fEnergy_AMeV(R3BDistribution1D::Delta(500))
+    , fPDGCode(2212)
+    , fA(1)
+    , fMass_GeV(ProtonMass)
+    , fRNG(0)
+{
+}
+
+R3BParticleGenerator::R3BParticleGenerator(Int_t pdgCode, unsigned int seed)
+    : fVertex_cm(R3BDistribution3D::Delta(0, 0, 0))
+    , fSpread_mRad({ R3BDistribution1D::Delta(0), R3BDistribution1D::Flat(0, 2 * TMath::Pi() * 1e3) })
+    , fEnergy_AMeV(R3BDistribution1D::Delta(500))
+    , fPDGCode(pdgCode)
+    , fA(1)
+    , fRNG(seed)
+{
+    SetParticle(pdgCode);
+}
+
+R3BParticleGenerator::R3BParticleGenerator(const Char_t* ionName, unsigned int seed)
+    : fVertex_cm(R3BDistribution3D::Delta(0, 0, 0))
+    , fSpread_mRad({ R3BDistribution1D::Delta(0), R3BDistribution1D::Flat(0, 2 * TMath::Pi() * 1e3) })
+    , fEnergy_AMeV(R3BDistribution1D::Delta(500))
+    , fRNG(seed)
+{
+    SetIon(ionName);
+}
+
+void R3BParticleGenerator::SetProton()
+{
+    fPDGCode = 2212;
+    fA = 1;
+    fMass_GeV = ProtonMass;
+};
+
+void R3BParticleGenerator::SetNeutron()
+{
+    fPDGCode = 2112;
+    fA = 1;
+    fMass_GeV = NeutronMass;
+};
+
+void R3BParticleGenerator::SetIon(const Char_t* ionName)
+{
+    auto ion = (FairIon*)FairRunSim::Instance()->GetUserDefIons()->FindObject(ionName);
+    if (!ion)
+        LOG(fatal) << "R3BParticleGenerator:  Ion is not defined!";
+
+    fPDGCode = ChargeMass2PDGCode(ion->GetZ(), ion->GetA());
+    fMass_GeV = ion->GetMass();
+};
+
+void R3BParticleGenerator::SetParticle(Int_t pdgCode)
+{
+    fPDGCode = pdgCode;
+    auto particle = TDatabasePDG::Instance()->GetParticle(pdgCode);
+    if (!particle)
+        LOG(fatal) << "R3BParticleGenerator: Unknown PDG Code!";
+
+    if (abs(pdgCode) > 1000000000)
```

Comment:
I would recommend not getting the masses for Ions from TDatabasePDG, but from Geant4, [like so](https://github.com/R3BRootGroup/R3BRoot/blob/8affc0f5150f0b8927f5c11a66cf29ceba0fed42/r3bgen/R3BPhaseSpaceGenerator.cxx#L52)
Also compare with the `AddHeavyIon` method, its been a while but `run->AddNewIon` might be necessary for the VMC layers to work with heavy ions.

Also, the SetIon/Particle interfaces differ between the different Generators - that might be a target to be unified.

Reply:
> I would recommend not getting the masses for Ions from TDatabasePDG, but from Geant4, [like so](https://github.com/R3BRootGroup/R3BRoot/blob/8affc0f5150f0b8927f5c11a66cf29ceba0fed42/r3bgen/R3BPhaseSpaceGenerator.cxx#L52)

Does is not somewhat work against the abstraction?

> Also compare with the AddHeavyIon method, its been a while but run->AddNewIon might be necessary for the VMC layers to work with heavy ions.

It seems like its only a list of userdefined ions?

> Also, the SetIon/Particle interfaces differ between the different Generators - that might be a target to be unified.

Yes I agree. I would actually prefer to keep the number of generators small, but because of the (at least for me) special way R3BIonGenerator works, we have to decide if we want to break backwards compatibility or add a new very simple generator (this one basically).

Reply:
> Does is not somewhat work against the abstraction?

Yes, it does. But the abstraction is, uhm, "not great" anyhow.  It has been a while, but I had problems with the masses from ROOT - give it a try and compare the values. Note that in principle, Geant4 has all the masses already and also uses those (I think), so specifying them again (different source of truth) will only lead to tears - similar reason why the masses must not be specified in the ascii-input files. 

Personally, I would cut as much of the abstraction layer as possible and just use Geant4.


> It seems like its only a list of user defined ions?

Yes. Why that lists exists and especially, why it needs (might be needed?) to be populated manually, I don't know.

Reply:
> Yes, it does. But the abstraction is, uhm, "not great" anyhow. It has been a while, but I had problems with the masses from ROOT - give it a try and compare the values. Note that in principle, Geant4 has all the masses already and also uses those (I think), so specifying them again (different source of truth) will only lead to tears - similar reason why the masses must not be specified in the ascii-input files.

Well FairRoot will use TDatabasePDG to get the masses if you pass the corresponding flag.
I feel like, that this might be a problem, but can (or rather should) not be solved here.

> Yes. Why that lists exists and especially, why it needs (might be needed?) to be populated manually, I don't know.

I added the lines. The next push will have them included.